feat(fauna): update adapter for FQL v10

employee451 opened the pull request:
This PR tries to do the same as https://github.com/nextauthjs/next-auth/pull/8403. Rather than make a PR to @linzjax's branch, i thought it would be easier to open a new PR.

## ☕️ Reasoning

Fauna's new query language, FQL v10, has been out for a while. This PR updates `adapter-fauna` to use FQL v10 and the new conventions that have come from Fauna's side.

### Breakdown of changes:

- Use singular, uppercase names for collections (this is the new naming schema Fauna uses [in their examples](https://docs.fauna.com/fauna/current/reference/fsl_reference/syntax))
- Stop using `fauna-schema-migrate`, and instead use the Fauna CLI and [FSL](https://docs.fauna.com/fauna/current/reference/fsl_reference/syntax) for defining the collections and indexes
- Switch from the `faunadb` to the `fauna` package and refactor all queries from FQL v4 to FQL v10
- Update the documentation and add a "Migrating from v1" section

## 🧢 Checklist

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

## 🎫 Affected issues

Fixes: https://github.com/nextauthjs/next-auth/issues/8394

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

employee451 committed:
feat(adapters): update adapter-fauna to work with fql v10

employee451 committed:
Merge branch 'main' into fauna-fql-v10

employee451 committed:
refactor: remove unused imports

employee451 committed:
fix(adapters): fix vitest command call in test.sh script

employee451 committed:
feat(adapters): only wait 3 seconds for db to start in test.sh file

employee451 committed:
Merge branch 'main' into fauna-fql-v10

employee451 committed:
Merge branch 'nextauthjs:main' into fauna-fql-v10

employee451 committed:
Merge branch 'main' into fauna-fql-v10

employee451 commented:
Hi @ndom91 & @balazsorban44. It would be amazing if we could get this merged. We are currently migrating our app to `next-auth` and we want to use the new naming conventions Fauna uses for collections. The tests are all passing🌞
Thanks for your work on Auth.js!

employee451 committed:
feat: merge branch 'main' into fauna-fql-v10

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Merge branch 'main' into fauna-fql-v10

balazsorban44 committed:
cleanup

balazsorban44 committed:
fix test runs

balazsorban44 commented on packages/adapter-fauna/src/index.ts:
```diff
@@ -10,140 +9,68 @@
+export type FaunaUser = ToFauna<AdapterUser>
+export type FaunaSession = ToFauna<AdapterSession>
+export type FaunaVerificationToken = ToFauna<VerificationToken>
+export type FaunaAccount = ToFauna<AdapterAccount> & any // TODO: Remove `& any`
```
I could not drop the any here. Any ideas why?

employee451 replied:
<img width="697" alt="Skærmbillede 2024-02-05 kl  10 37 15" src="https://github.com/nextauthjs/next-auth/assets/22524988/34fa3f89-12a1-469c-9710-afb6eacd2dd0">

It has to do with the last line in this interface (`readonly [parameter: string]: JsonValue | undefined;`). 
Even if I change it to be `readonly [parameter: string]: string;`, it still doesn't work (even though that should be compatible with Fauna's `QueryValue` type (`undefined` is not a valid type for `QueryValue`). If I remove the last line, however, the type error disappears.

ndom91 replied:
@employee451 could you take another look at this todo? Would be great to get that `any` out of there :pray:

employee451 replied:
@ndom91 I have just pushed a fix for this. Please take a look

balazsorban44 commented on packages/adapter-fauna/src/index.ts:
```diff
@@ -10,140 +9,68 @@
+email,
+name,
+image,
+id,
+emailVerified
+`
 
-      throw error
-    }
-  }
-}
+export const sessionFields = fql`
+sessionToken,
+userId,
+expires
+`
+
+export const verificationTokenFields = fql`
+identifier,
+expires,
+token
+`
```
This should not be hardcoded. Any ideas on how to achieve this? The fields should be driven by the object keys of the argument of the given adapter method. Maybe `Object.keys(obj).join(",\n")` or something similar can help here.

employee451 replied:
Fixed in https://github.com/nextauthjs/next-auth/pull/9849/commits/93fb4531ffe021215b4f4236ad03c70647d9e6ef

balazsorban44 reviewed (changes requested):
Unfortunately tests did _not_ pass. <picture data-single-emoji=":thinking:" title=":thinking:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/thinking/a89918a57b536d5b.gif" alt=":thinking:" align="absmiddle"></picture> I tried fixing them (missing `fauna-shell`, too short `sleep`), but I get:

```sh
test/index.test.ts > createUser
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > getUser
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > getUserByEmail
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > createSession
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > getSessionAndUser
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > updateUser
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > updateSession
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > linkAccount
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > getUserByAccount
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > deleteSession
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > createVerificationToken
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > useVerificationToken
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > unlinkAccount
     → Response is in an unkown format: [object Object]
   ❯ test/index.test.ts > deleteUser
     → Response is in an unkown format: [object Object]
```

(Looks like the error message of `fauna` has a typo :grimacing:, `unkown` -> `unknown` )

Could you have a look and make the tests pass?

I also refactored the code as described in https://authjs.dev/guides/adapters/creating-a-database-adapter#official-adapter-guidelines

Most importantly, adapter methods should not assume any field names so it remains flexible.

Formatting should happen the way in and out of the database too. Auth.js assumes plain JS values.

balazsorban44 committed:
update lockfile

employee451 committed:
fix: make tests pass

employee451 committed:
fix(adapters): use id function in adapter tests

employee451 committed:
refactor: stop specifying field names

employee451 commented on packages/utils/adapter.ts:
```diff
@@ -65,7 +65,7 @@ export async function runBasicTests(options: TestOptions) {
 
   let user = options.fixtures?.user ?? {
-    id: randomUUID(),
+    id: id(),
```
@balazsorban44 I made these changes in this file. We needed to specify the custom function for generating ID's (see `packages/adapter-fauna/test/index.test.ts` in this PR). But the custom function an adapter could specify was only used once in this file and not in the other cases which I changed here

employee451 replied:
Should I open a new PR for this change?

employee451 committed:
docs: remove data.id unique constraint on user collection in docs

employee451 committed:
feat: merge branch 'main' into fauna-fql-v10

employee451 committed:
Merge branch 'main' into fauna-fql-v10

employee451 committed:
chore: make major version bump

employee451 commented on packages/adapter-fauna/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@auth/fauna-adapter",
-  "version": "1.3.2",
+  "version": "2.0.0",
```
As discussed in this comment https://github.com/nextauthjs/next-auth/issues/8394#issuecomment-1690655534, we should make a major version bump for the package. People who previously used the adapter need to rename their Fauna collections to make 2.0 work. 
I also considered accepting a config in the adapter, specifying what the collection names are. These could then be used with string interpolation, so people can use any naming schema they want for their collections (or keep the old ones). Let me know what you think.

employee451 replied:
If you think the collection names config sounds like a good idea, I'd like to open a new PR for this after we get this merged

employee451 committed:
Merge branch 'nextauthjs:main' into fauna-fql-v10

employee451 committed:
Merge branch 'main' into fauna-fql-v10

employee451 committed:
Merge branch 'nextauthjs:main' into fauna-fql-v10

employee451 committed:
Merge branch 'main' into fauna-fql-v10

employee451 committed:
Merge branch 'main' into fauna-fql-v10

employee451 committed:
Merge branch 'main' into fauna-fql-v10

employee451 committed:
Merge branch 'main' into fauna-fql-v10

employee451 committed:
Merge branch 'main' into fauna-fql-v10

employee451 commented:
Sorry if I'm being pushy. We really want to get this merged so we can switch from Auth0 to next-auth with an email provider. I've made the changes you requested and the tests are passing properly now ✅ 🌞
@balazsorban44 @ndom91 @ThangHuuVu

employee451 committed:
Merge branch 'main' into fauna-fql-v10

ndom91 commented:
Hey @employee451 first of all thanks for the contribution! 

Can you give me a short run down of the implications of this for existing fauna adapter users?

employee451 commented:
> Hey @employee451 first of all thanks for the contribution!
> 
> Can you give me a short run down of the implications of this for existing fauna adapter users?

Thanks for your reply @ndom91!
Existing Fauna adapter users will need to use a Fauna client from the new `fauna` library instead of the old `faunadb`. Additionally, they will have to rename the auth collections that the adapter uses. This is explained in the new "Migrating from v1" section I've added to the adapter docs:

```
In v2, we've renamed the collections to use uppercase naming, in accordance with Fauna best practices. If you're migrating from v1, you'll need to rename your collections to match the new naming scheme.
Additionally, we've renamed the indexes to match the new method-like index names. Please see above for the new index definitions.
```

employee451 committed:
Merge branch 'main' into fauna-fql-v10

employee451 committed:
Merge branch 'main' into fauna-fql-v10

employee451 committed:
fix: remove use of "& any"

employee451 committed:
Merge branch 'main' into fauna-fql-v10

ndom91 reviewed (approved):
LGTM :tada:

ndom91 committed:
Update packages/adapter-fauna/src/index.ts

employee451 commented:
> LGTM 🎉

Thanks! What needs to happen to get this merged? Do we need @balazsorban44 to approve the changes?

ndom91 commented:
@employee451 what do you think about this migration script in FQL? Looks like it should do the trick to me:

```fauna
Collection.byName("accounts")!.update({
  name: "Account"
  indexes: {
    byUserId: {
        terms: [{ field: "userId" }]
    },
    byProviderAndProviderAccountId: {
        terms: [{ field: "provider" }, { field: "providerAccountId" }]
    }
  }
})
Collection.byName("sessions")!.update({
  name: "Session",
  indexes: {
    bySessionToken: {
        terms: [{ field: "sessionToken" }]
    },
    byUserId: {
        terms: [{ field: "userId" }]
    },
  }
})
Collection.byName("users")!.update({
  name: "User",
  indexes: {
    byEmail: {
        terms: [{ field: "email" }]
    },
  }
})
Collection.byName("verification_tokens")!.update({
  name: "VerificationToken",
  indexes: {
    byIdentifierAndToken: {
        terms: [{ field: "identifier" }, { field: "token" }]
    },
  }
})
```

I can merge it, but we'll need someone else to cut a release. We plan on doign that this weekend anyway. No need to tag anyone else :wink: :+1:

employee451 commented:
> @employee451 what do you think about this migration script in FQL? Looks like it should do the trick to me:
> 
> ```
> Collection.byName("accounts")!.update({
>   name: "Account"
>   indexes: {
>     byUserId: {
>         terms: [{ field: "userId" }]
>     },
>     byProviderAndProviderAccountId: {
>         terms: [{ field: "provider" }, { field: "providerAccountId" }]
>     }
>   }
> })
> Collection.byName("sessions")!.update({
>   name: "Session",
>   indexes: {
>     bySessionToken: {
>         terms: [{ field: "sessionToken" }]
>     },
>     byUserId: {
>         terms: [{ field: "userId" }]
>     },
>   }
> })
> Collection.byName("users")!.update({
>   name: "User",
>   indexes: {
>     byEmail: {
>         terms: [{ field: "email" }]
>     },
>   }
> })
> Collection.byName("verification_tokens")!.update({
>   name: "VerificationToken",
>   indexes: {
>     byIdentifierAndToken: {
>         terms: [{ field: "identifier" }, { field: "token" }]
>     },
>   }
> })
> ```
> 
> I can merge it, but we'll need someone else to cut a release. We plan on doign that this weekend anyway. No need to tag anyone else 😉 👍

The script looks good😄

ndom91 commented:
Ah forgot to drop the existing indexes, here's a slightly updated one: 

```ts
Collection.byName("accounts")!.update({
  name: "Account"
  indexes: {
    byUserId: {
        terms: [{ field: "userId" }]
    },
    byProviderAndProviderAccountId: {
        terms: [{ field: "provider" }, { field: "providerAccountId" }]
    },
    account_by_provider_and_provider_account_id: null,
    accounts_by_user_id: null
  }
})
Collection.byName("sessions")!.update({
  name: "Session",
  indexes: {
    bySessionToken: {
        terms: [{ field: "sessionToken" }]
    },
    byUserId: {
        terms: [{ field: "userId" }]
    },
    session_by_session_token: null,
    sessions_by_user_id: null
  }
})
Collection.byName("users")!.update({
  name: "User",
  indexes: {
    byEmail: {
        terms: [{ field: "email" }]
    },
    user_by_email: null
  }
})
Collection.byName("verification_tokens")!.update({
  name: "VerificationToken",
  indexes: {
    byIdentifierAndToken: {
        terms: [{ field: "identifier" }, { field: "token" }]
    },
    verification_token_by_identifier_and_token: null
  }
})
```

If you also think its good, lets include it in the docs. What do you think?

ndom91 commented:
Awesome, if you cuold add that to this PR, i'll approve it again :+1:

employee451 committed:
feat: add migration script to docs

ndom91 commented:
Okay merged it, thanks again! Like I said, we plan on getting a release out this weekend, so I'll keep you posted :+1:

employee451 commented:
Thank you so much! I'm excited for v5🍃

ndom91 commented:
@employee451 available in `@auth/fauna-adapter@2.1.0` now :tada:
